readme.txt
Network printer setup utility (netprinters.exe)

This utility is aimed at creation of networked printers on windows network
clients, it should run on windows 2000 or newer. Compiling the source requires
MS visual studio and MinGW, a pre-compiled binary is included if you don't want
to create your own build.

The configuration file format is: "<command> <machine> [<printer>]"
Printer names may include whitespace, both tabs and spaces may be used as
seperators and lines beginning with a "#" are ignored as comments.

The available commands are:

printer	- Add a networked printer by UNC path
default	- Set a default printer by UNC path
exiton	- Exit the program

Commands are only executed if the <machine> expression matches the hostname of
the local machine, hostnames may contain the * and ? wildcards. The printer and
default commands obviously require a printer name, exiton doesn't.

You can use netprinters.exe in a login script with the following code, assuming
your configuration file is named printers.txt and is in the root of NETLOGON.

NET USE L: %logonserver%\NETLOGON /persistent:no
L:\netprinters.exe L:\printers.txt
net use L: /DELETE

In order to make a custom build of netprinters open the Visual Studio command
line, cd to this directory and run build.bat


CHANGES
NetPrinters changelog

Version 2.1:
	Wrote new expression comparing function with support for a '#' wildcard
	which is the same as '?', except it only matches integers.
	
	Removed -a argument because it didn't have any useful information.
	Removed -e argument because it was only there for debugging anyway.
	
	Wrote ncase_match() function for case-insensitive string comparisons.

Version 2.0:
	Full rewrite of v1.x


src/compare.c
/* Flexible string comparison function with wildcard support
 * Version 1.23 (21/12/2007)
 * Copyright (C) 2007 Daniel Collins <solemnwarning@solemnwarning.net>
 *
 * Released to public domain, I take no responsibility for this code or it's
 * users, it may work, it may not, it may destroy the universe but it has no
 * warranty, so no sending lawyers my way if this code breaks somthing.
 *
 * If you find a bug please email me :)
*/

/* Version history:
 *
 * 1.0 (11/06/2007):
 * Initial release
 *
 * 1.1 (11/06/2007):
 * Fixed a bug:
 *	maxlen was read with stdarg, whether the STR_MAXLEN flag was set or
 *	not, this causes undefined behavior according to C so it could crash
 *	or cause other errors on some platforms.
 *
 * 1.2 (23/08/2007):
 * Rewrote to be more flexible
 * Fixed a bug:
 *	The * wildcard between two characters in one string caused a mismatch
 *	when the other string has 0 characters in its place because I used ++
 *	where I should have used += 2.
 *
 * 1.21 (09/09/2007):
 * Created a header for this function (compare.h)
 *
 * 1.22 (06/11/2007):
 * Added code to make str_compare() handle NULL pointers as strings.
 *	(It treats both NULL as match, one NULL as mismatch)
 *
 * 1.23 (21/12/2007):
 * Fixed a bug:
 *	If STR_MAXLEN was used and a maxlen of zero was passed, str_compare()
 *	would act as if STR_MAXLEN had not been used.
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <ctype.h>
#include <string.h>

#include "compare.h"

/* Compare str1 and str2
 * Returns 1 on match, 0 on mismatch
*/
int str_compare(char const* str1, char const* str2, int flags, ...) {
	if(str1 == NULL && str2 == NULL) {
		return(1);
	}
	if(str1 == NULL || str2 == NULL) {
		return(0);
	}
	
	size_t maxlen = 0;
	size_t compared = 0;
	
	va_list arglist;
	va_start(arglist, flags);
	if(flags & STR_MAXLEN) {
		maxlen = va_arg(arglist, size_t);
	}
	va_end(arglist);
	
	for(; !(flags & STR_MAXLEN) || compared < maxlen; compared++) {
		if(str1[0] == str2[0]) {
			if(str1[0] == '\0') {
				break;
			}
			
			str1++;
			str2++;
			continue;
		}
		if(flags & STR_NOCASE && (tolower(str1[0]) == tolower(str2[0]))) {
			str1++;
			str2++;
			continue;
		}
		if((flags & STR_WILDCARDS || flags & STR_WILDCARD1) && (str1[0] == '*' || str1[0] == '?')) {
			if(str1[0] == '?' && str2[0] != '\0') {
				str1++;
				str2++;
				continue;
			}
			if(str1[0] == '*') {
				if(str2[0] == '\0' || str1[1] == '\0') {
					break;
				}
				if(str1[1] == str2[0]) {
					str1 += 2;
				}
				str2++;
				continue;
			}
		}
		if((flags & STR_WILDCARDS || flags & STR_WILDCARD2) && (str2[0] == '*' || str2[0] == '?')) {
			if(str2[0] == '?' && str1[0] != '\0') {
				str1++;
				str2++;
				continue;
			}
			if(str2[0] == '*') {
				if(str1[0] == '\0' || str2[1] == '\0') {
					break;
				}
				if(str2[1] == str1[0]) {
					str2 += 2;
				}
				str1++;
				continue;
			}
		}
		
		return(0);
	}
	return(1);
}


src/compare.h
/* Flexible string comparison function with wildcard support - header
 * Version 1.23 (21/12/2007)
 * Copyright (C) 2007 Daniel Collins <solemnwarning@solemnwarning.net>
 *
 * Released to public domain, I take no responsibility for this code or it's
 * users, it may work, it may not, it may destroy the universe but it has no
 * warranty, so no sending lawyers my way if this code breaks somthing.
 *
 * If you find a bug please email me :)
*/

/* Read compare.c for more information */

#ifndef SOLEMN_COMPARE_H
#define SOLEMN_COMPARE_H

/* Flags, bitwise OR them */
#define STR_NOCASE	1	/* Do case-insensitive comparision */
#define STR_MAXLEN	2	/* Don't compare more then ... bytes */
#define STR_WILDCARDS	4	/* Parse wildcard characters * and ? */
#define STR_WILDCARD1	8	/* Parse wildcard characters * and ? in str1 */
#define STR_WILDCARD2	16	/* Parse wildcard characters * and ? in str2 */

/* Compare str1 and str2
 * Returns 1 on match, 0 on mismatch
*/
int str_compare(char const*, char const*, int, ...);

#endif /* !SOLEMN_COMPARE_H */


src/printer.c
/* NetPrinters - Printer functions
 * Copyright (C) 2008 Daniel Collins <solemnwarning@solemnwarning.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the distribution.
 *
 *	* Neither the name of the author nor the names of its contributors may
 *	  be used to endorse or promote products derived from this software
 *	  without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <windows.h>
#include <stdio.h>
#include <stdlib.h>

#define EPRINTF(...) fprintf(stderr, __VA_ARGS__)

/* Like the perror() function, but uses windows's backwards API to get error
 * messages
*/
static char* np_errmsg(void) {
	static char errmsg[1024] = {'\0'};
	FormatMessage(
		FORMAT_MESSAGE_FROM_SYSTEM,
		NULL,
		GetLastError(),
		0,
		errmsg,
		1023,
		NULL
	);
	errmsg[strcspn(errmsg, "\n\r")] = '\0';
	
	return errmsg;
}

/* Add a printer connection by UNC path */
void printer_add(char* uncpath) {
	if(!AddPrinterConnection(uncpath)) {
		EPRINTF("Can't add printer %s: %s\n", uncpath, np_errmsg());
	}else{
		printf("Added printer:\t\t%s\n", uncpath);
	}
}

/* Set default printer by UNC path */
void printer_default(char* uncpath) {
	if(!SetDefaultPrinter(uncpath)) {
		EPRINTF("Can't set default printer %s: %s\n", uncpath, np_errmsg());
	}else{
		printf("Set default printer:\t%s\n", uncpath);
	}
}


src/netprinters.c
/* NetPrinters - Source
 * Copyright (C) 2008 Daniel Collins <solemnwarning@solemnwarning.net>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *	* Redistributions of source code must retain the above copyright
 *	  notice, this list of conditions and the following disclaimer.
 *
 *	* Redistributions in binary form must reproduce the above copyright
 *	  notice, this list of conditions and the following disclaimer in the
 *	  documentation and/or other materials provided with the distribution.
 *
 *	* Neither the name of the author nor the names of its contributors may
 *	  be used to endorse or promote products derived from this software
 *	  without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <windows.h>
#include <winsock2.h>
#include <lm.h>
#include <security.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define VERSION "v2.1"
#define WHITESPACE "\r\n\t "

#define EPRINTF(...) fprintf(stderr, __VA_ARGS__)
#define ARGN_IS(arg) (strcmp(argv[argn], arg) == 0)
#define fprintf(fh, ...) fprintf(fh, __VA_ARGS__); fflush(fh);
#define printf(...) printf(__VA_ARGS__); fflush(stdout);

static void print_usage(void);
static char **get_printers(void);
static char *win32_strerr(DWORD errnum);
static void list_printers(void);
static void connect_printer(char *printer);
static void default_printer(char *printer);
static void disconnect_printer(char *printer);
static void disconnect_by_expr(char *expr);
static void exec_script(char const *filename);
static void load_env(void);
static int expr_compare(char const *str, char const *expr);
static int ncase_match(char const *str1, char const *str2);

static struct {
	char username[1024];
	char nbname[1024];
} userenv = {{'\0'},{'\0'}};

static void print_usage(void) {
	printf("Usage: netprinters.exe <arguments>\n");
	printf("Arguments:\n\n");
	
	printf("-c <UNC path>\tConnect to a printer\n");
	printf("-d <UNC path>\tSet default printer\n");
	printf("-r <expression>\tDelete any matching printer connections\n");
	printf("-l\t\tList connected printers\n");
	printf("-s <filename>\tExecute a netprinters script\n");
}

/* Returns a NULL-terminated list of connected printers obtained from the
 * EnumPrinters() win32 call, or NULL on error.
*/
static char **get_printers(void) {
	PRINTER_INFO_4 *printers = NULL;
	DWORD size = 0, count, n;
	char **retbuf = NULL;
	
	while(!EnumPrinters(PRINTER_ENUM_CONNECTIONS, NULL, 4, (void*)printers, size, &size, &count)) {
		if(GetLastError() != 122 && GetLastError() != 1784) {
			EPRINTF("Can't fetch printers: %s\n", win32_strerr(GetLastError()));
			goto GET_PRINTERS_END;
		}
		
		free(printers);
		if((printers = malloc(size)) == NULL) {
			EPRINTF("Can't allocate %u bytes\n", (unsigned int)size);
			goto GET_PRINTERS_END;
		}
	}
	
	if(!(retbuf = malloc(sizeof(char*) * (count+1)))) {
		EPRINTF("Can't allocate %d bytes\n", (int)(sizeof(char*) * (count+1)));
		goto GET_PRINTERS_END;
	}
	retbuf[count] = NULL;
	
	for(n = 0; n < count; n++) {
		char *pname = printers[n].pPrinterName;
		
		if((retbuf[n] = malloc(strlen(pname)+1)) == NULL) {
			EPRINTF("Can't allocate %u bytes\n", (strlen(pname)+1));
			break;
		}
		strcpy(retbuf[n], pname);
	}
	
	GET_PRINTERS_END:
	free(printers);
	return retbuf;
}

/* Equvilent of the strerr() function, using windows's backwards FormatMessage
 * API call.
 *
 * Returns string stored in static buffer
*/
static char *win32_strerr(DWORD errnum) {
	static char buf[1024] = {'\0'};
	
	FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM, NULL, errnum, 0, buf, 1023, NULL);
	buf[strcspn(buf, "\r\n")] = '\0';
	return buf;	
}

/* List printers to stdout */
static void list_printers(void) {
	char **printers = get_printers();
	unsigned int pnum = 0;
	
	while(printers[pnum]) {
		puts(printers[pnum]);
		free(printers[pnum]);
		
		pnum++;
	}
	
	free(printers);
}

/* Connect to a network printer */
static void connect_printer(char *printer) {
	if(AddPrinterConnection((char*)printer)) {
		printf("Added printer:\t\t%s\n", printer);
	}else{
		EPRINTF(
			"Can't connect to printer %s: %s\n",
			printer, win32_strerr(GetLastError())
		);
	}
}

/* Set default printer */
static void default_printer(char *printer) {
	if(SetDefaultPrinter(printer)) {
		printf("Set default printer:\t%s\n", printer);
	}else{
		EPRINTF(
			"Can't make printer %s default: %s\n",
			printer, win32_strerr(GetLastError())
		);
	}
}

/* Disconnect from a printer */
static void disconnect_printer(char *printer) {
	if(DeletePrinterConnection(printer)) {
		printf("Disconnected from:\t%s\n", printer);
	}else{
		EPRINTF(
			"Can't disconnect from printer %s: %s\n",
			printer, win32_strerr(GetLastError())
		);
	}
}

/* Disconnect from any printers matching the supplied expression */
static void disconnect_by_expr(char *expr) {
	char **printers = get_printers();
	unsigned int pnum = 0;
	
	while(printers[pnum]) {
		char *pname = printers[pnum++];
		
		if(expr_compare(pname, expr)) {
			disconnect_printer(pname);
		}
		
		free(pname);
	}
	
	free(printers);
}

/* Parse and execute a NetPrinters script */
static void exec_script(char const *filename) {
	FILE *fh = fopen(filename, "r");
	if(!fh) {
		EPRINTF(
			"Can't open script %s: %s\n", filename,
			win32_strerr(GetLastError())
		);
		return;
	}
	
	char buf[1024];
	unsigned int lnum = 1;
	int sblock = 0;
	
	while(fgets(buf, 1024, fh)) {
		char *name = buf+strspn(buf, WHITESPACE);
		char *value = name+strcspn(name, WHITESPACE);
		
		if(value[0] != '\0') {
			value[0] = '\0';
			value++;
			value += strspn(value, WHITESPACE);
			value[strcspn(value, "\r\n")] = '\0';
		}
		
		if(name[0] == '\0') {
			sblock = 0;
		}
		if(name[0] == '#' || name[0] == '\0' || sblock) {
			lnum++;
			continue;
		}
		
		if(ncase_match(name, "AddPrinter")) {
			connect_printer(value);
		}else if(ncase_match(name, "DefaultPrinter")) {
			default_printer(value);
		}else if(ncase_match(name, "DeletePrinter")) {
			disconnect_by_expr(value);
		}else if(ncase_match(name, "Exit")) {
			printf("Line %u:\tExit used\n", lnum);
			exit(0);
		}else if(ncase_match(name, "NetBIOS")) {
			if(!expr_compare(userenv.nbname, value)) {
				sblock = 1;
			}
		}else if(ncase_match(name, "!NetBIOS")) {
			if(expr_compare(userenv.nbname, value)) {
				sblock = 1;
			}
		}else if(ncase_match(name, "Username")) {
			if(!expr_compare(userenv.username, value)) {
				sblock = 1;
			}
		}else if(ncase_match(name, "!Username")) {
			if(expr_compare(userenv.username, value)) {
				sblock = 1;
			}
		}else{
			EPRINTF(
				"Unknown directive %s at line %u\n",
				name, lnum
			);
		}
		
		lnum++;
	}
	
	if(fclose(fh) != 0) {
		EPRINTF(
			"Can't close script %s: %s\n", filename,
			win32_strerr(GetLastError())
		);
	}
}

/* Read the environment information into the userenv structure */
static void load_env(void) {
	DWORD bsize;
	
	bsize = sizeof(userenv.nbname)-1;
	GetComputerName(userenv.nbname, &bsize);
	
	bsize = sizeof(userenv.username)-1;
	GetUserName(userenv.username, &bsize);
}

/* Compare the supplied string and expression
 * Returns 1 upon match, zero otherwise.
*/
static int expr_compare(char const *str, char const *expr) {
	while(1) {
		if(expr[0] == '\0' && str[0] != '\0') {
			return 0;
		}
		if(str[0] == '\0') {
			if(expr[0] == '\0' || expr[0] == '*') {
				break;
			}
			
			return 0;
		}
		
		if(expr[0] == '*') {
			if(expr[1] == str[0]) {
				expr += 2;
			}
			str++;
			
			continue;
		}
		if(expr[0] == '?' && str[0] != '\0') {
			expr++;
			str++;
			
			continue;
		}
		if(expr[0] == '#' && isdigit(str[0])) {
			expr++;
			str++;
			
			continue;
		}
		if(tolower(expr[0]) == tolower(str[0])) {
			expr++;
			str++;
			
			continue;
		}
		
		return 0;
	}
	
	return 1;
}

/* Compare two strings, ignoring case
 * Returns 1 if they match, zero otherwise.
*/
static int ncase_match(char const *str1, char const *str2) {
	size_t pos = 0;
	
	while(tolower(str1[pos]) == tolower(str2[pos])) {
		if(str1[pos] == '\0') {
			return 1;
		}
		
		pos++;
	}
	
	return 0;
}

int main(int argc, char** argv) {
	if(argc < 2) {
		print_usage();
		return 1;
	}
	
	printf("NetPrinters " VERSION "\n");
	printf("Copyright (C) 2008 Daniel Collins\n\n");
	
    	if(LOBYTE(LOWORD(GetVersion())) < 5) {
		EPRINTF("This program requires Windows 2000 or newer\n");
		return 1;
	}
	
	load_env();
	
	int argn = 1;
	while(argn < argc) {
		if(ARGN_IS("-c")) {
			if((argn + 1) == argc) {
				EPRINTF("-c requires an argument");
				return 1;
			}
			
			connect_printer(argv[++argn]);
		}else if(ARGN_IS("-d")) {
			if((argn + 1) == argc) {
				EPRINTF("-d requires an argument");
				return 1;
			}
			
			default_printer(argv[++argn]);
		}else if(ARGN_IS("-r")) {
			if((argn + 1) == argc) {
				EPRINTF("-r requires an argument");
				return 1;
			}
			
			disconnect_by_expr(argv[++argn]);
		}else if(ARGN_IS("-l")) {
			list_printers();
			return 0;
		}else if(ARGN_IS("-s")) {
			if((argn + 1) == argc) {
				EPRINTF("-s requires an argument");
				return 1;
			}
			
			exec_script(argv[++argn]);
		}else{
			EPRINTF("Unknown argument: %s\n", argv[argn]);
			return 1;
		}
		
		argn++;
	}
	
	return 0;
}
